include/glow/Runtime/Executor/NetworkExecutionState.h

```diff
@@ -0,0 +1,156 @@
+  explicit NetworkExecutionState(const DAGNode *root);
+
+  /// Destructor.
+  ~NetworkExecutionState();
+
+  /// Does the BFS traversal and initializes the NetworkExecutionState.
+  void init(const DeviceManagerMapTy &devices);
+
+  /// Binds the state to a new run. This moves the result ctx and cb to be owned
+  /// by the networkExecutionState for the duration of the run.
+  void bind(std::unique_ptr<ExecutionContext> resultCtx, ResultCBTy cb,
+            RunIdentifierTy runId);
+
+  /// \returns a unique pointer to an input bindings for \p node. This should
+  /// not be called at the same time as insertIntoNodeCtx().
+  std::unique_ptr<ExecutionContext>
+  getUniqueNodeContextPtr(const DAGNode *node);
+
+  /// Returns the intermediateContext back to the networkExecutionState after
+  /// completion so it can be re-used.
+  void returnUniqueNodeContextPtr(const DAGNode *node,
+                                  std::unique_ptr<ExecutionContext> ctx);
+
+  /// Increment the count of inflight nodes by \p increment (default is 1).
+  void incrementInflightNodes(unsigned increment = 1);
+
+  /// Decrement the count of inflight nodes by the \p decrement (default is 1).
+  /// \returns true if there are no nodes inflight after the decrement
+  /// operation.
+  bool decrementInflightNodes(unsigned decrement = 1);
+
+  /// Increment the count of completed parent nodes for \p node. \returns
+  /// true if all parents are done after the increment operation, false
+  /// otherwise.
+  bool incrementNodeParentsDone(const DAGNode *node, unsigned increment = 1);
+
+  /// Move all events from the provided vector into the top level resultContxt.
+  void insertIntoTraceContext(TraceContext *runCtx);
+
+  /// \returns a unique pointer to the result bindings. This should not be
+  /// called at the same time as getRawResultPlaceholderBindingsPtr() or
+  /// insertIntoResultCtx().
+  std::unique_ptr<ExecutionContext> getUniqueResultContextPtr();
+
+  /// \returns a raw pointer to the result bindings. This should be not called
+  /// at the same time as getUniqueResultPlaceholderBindingsPtr().
+  ExecutionContext *getRawResultContextPtr() const;
+
+  /// \returns the callback for this execution.
+  ResultCBTy getCallback() { return cb_; }
+
+  /// \returns the OneErrOnly Error container for the execution.
+  OneErrOnly &getErrorContainer() { return errContainer_; }
+
+  /// \returns the run ID for the execution.
+  RunIdentifierTy getRunId() const { return runId_; }
+
+  /// \returns the Executor used for this run.
+  // ThreadExecutor *getExecutor() { return executor_; }
+
+  /// Whether or not this node has been initialized.
+  bool initialized_{false};
+
+private:
+  /// The run identifier for this execution of a DAG.
+  RunIdentifierTy runId_;
+  /// The callback that should be called when execution is done.
+  ResultCBTy cb_;
+
+  /// Map of all buffers allocated for intermediate contexts.
+  std::unordered_map<Placeholder *, void *> buffers_;
```

Comment:
Can we group these members by things that are static (things we can determine when we create the state pool except for the tensors that we need to bind), and things are dynamic (mainly bookkeeping of the dag job status).